agent/src/beerocks/slave/son_slave_thread.cpp

```diff
@@ -3862,77 +3862,47 @@ bool slave_thread::handle_autoconfiguration_renew(Socket *sd, ieee1905_1::CmduMe
-                return false;
-            }
-            auto m2 = std::dynamic_pointer_cast<WSC::m2>(WSC::AttrList::parse(*tlvWsc));
-            if (!m2) {
-                LOG(ERROR) << "Not an M2!";
-                return false;
-            }
-            m2_list.push_back(m2);
-        } else if (type == int(ieee1905_1::eTlvType::TLV_VENDOR_SPECIFIC)) {
-            // If this is an Intel Agent, it will have VS TLV as the last TLV.
-            // Currently, we don't support skipping TLVs, so if we see a VS TLV, we assume
-            // It is an Intel agent, and will add the class in the Intel agent handling below.
-            intel_agent = true;
-            break;
-        } else if (type == int(wfa_map::eTlvTypeMap::TLV_AP_RADIO_BASIC_CAPABILITIES)) {
-            // Check if this is a M1 message that we sent to the controller, which was just looped back.
-            // The M1 and M2 messages are both of CMDU type AP_Autoconfiguration_WSC. Thus,
-            // when we send the M2 to the local agent, it will be published back on the local bus because
-            // the destination is our AL-MAC, and the controller does listen to this CMDU.
-            // Ideally, we should use the message type attribute from the WSC payload to distinguish.
-            // Unfortunately, that is a bit complicated with the current tlv parser. However, there is another
-            // way to distinguish them: the M1 message has the AP_Radio_Basic_Capabilities TLV,
-            // while the M2 has the AP_Radio_Identifier TLV.
-            // If this is a looped back M1 CMDU, we can treat is as handled successfully.
-            LOG(DEBUG) << "Loopbed back M1 CMDU";
-            return true;
-        } else {
-            LOG(ERROR) << "Uknown TLV, type " << std::hex << type;
-            return false;
-        }
-        type = cmdu_rx.getNextTlvType();
-    }
-
+    auto ruid = cmdu_rx.getClass<wfa_map::tlvApRadioIdentifier>();
     if (!ruid) {
-        LOG(ERROR) << "Failed to get tlvApRadioIdentifier TLV";
+        LOG(ERROR) << "getClass<wfa_map::tlvApRadioIdentifier> failed";
         return false;
     }
-
     // Check if the message is for this radio agent by comparing the ruid
     if (network_utils::mac_from_string(config.radio_identifier) != ruid->radio_uid()) {
         LOG(DEBUG) << "not to me - ruid " << config.radio_identifier
                    << " != " << network_utils::mac_to_string(ruid->radio_uid());
         return true;
     }
 
+    std::list<WSC::m2> m2_list;
+    for (size_t i = 0; i < cmdu_rx.msg.getClassCount<ieee1905_1::tlvWsc>(); i++) {
+        auto tlvWsc = cmdu_rx.msg.getClass<ieee1905_1::tlvWsc>(i);
+        if (tlvWsc) {
+            auto m2 = std::dynamic_pointer_cast<WSC::m2>(tlvWsc->getInnerClassList());
+            if (m2)
+                m2_list.push_back(*m2);
+        }
+    }
     if (m2_list.empty()) {
         LOG(ERROR) << "No M2s present";
         return false;
     }
```

Comment:
In my opinion, instead of iterating over the cmdu_rx with `getClassCount<T>()`  we should implement a function named something like `getAllClasses<T>()` (`getClassList<T>()` would be too confusing) that returns a vector of all tlvs of type T in the cmdu.
- Firstly, it would be much more befitting of the OOP oriented direction of this pull request and it's predecessors.
- Secondly, once we have that, we could iterate over the vector with a foreach loop, making boilerplates such as this one shorter and more readable.

Reply:
Good idea - done.